### ISSUE TITLE
fixed speaker configuration support per device

### ISSUE THREAD

**Issue by Andy2244** [sirWest/AudioSwitch#7]
Hi,

it would be nice to-be able to set a fixed speaker configuration per device, that is set/used on every device switch. The main problem is that Win7 resets the speaker configuration of receivers, which prevents proper 5.1 output for games.

So while i can switch the device quickly using AudioSwitch, i still have to manually set the 5.1 speaker configuration every time my receiver was in standby.

I like the simplicity of your tool, so if i had the ability to set a default speaker configuration per audio device this would allow me to have the device switch + speaker set in a single click.

Example of the problem:
http://superuser.com/questions/695778/preventing-windows-7-from-reverting-to-stereo

In my case i frequently switch from stereo desktop mode on my Xi-Fi to a 5.1 HDMI device (receiver), but every time i resume my PC with the receiver still in sleep, windows will reset the HDMi configuration to stereo. If i than switch to TV gaming mode, many games read this configuration and only output stereo sound instead of 5.1. This is quite annoying and your tool could elegantly solve this issue.

thx
Andy

**Comment by sirWest** [comment]
This requires AudioSwitch to run with administrator privileges as describer everywhere, here for ex https://social.msdn.microsoft.com/Forums/windowsdesktop/en-US/e2e9858f-9bf5-4400-8a4e-570fa3285aad/changing-audio-device-sample-rate-via-pkeyaudioenginedeviceformat-does-not-work?forum=windowssdk
I don't think I see a point to implement this as this would require AudioSwitch to be running as admin all the time to monitor the speaker configuration and is an overkill for a tray application. In theory I could make a parallel version with this feature and requiring admin access with each launch. This would probably be for only few people though who really miss it.

**Comment by Andy2244** [comment]
Thats unfortunate, it basically forces me to use this tool, (http://blog.contriving.net/2009/05/04/a-hotkey-to-switch-between-headphones-and-speakers-soundswitch/) which is more complicated to setup/use.

The problem is even this tool has separate hotkeys for device and configuration sets, while all i want is switch to the HDMI device via a mouse-click and be sure the speaker configuration is also correctly set to 5.1.

The problem might be more widespread than assumed, since there is no easy fix and it affects any TV gamer with a A/V receiver.

**Comment by sirWest** [comment]
The only way without admin access is to do it the same way it's done everywhere, the very dirty way - using keypresses also. I could either add support for user-defined AHK script after device switch or program it into AudioSwitch. I will try it out and see how ugly it is and think about it then. If the logic is consistent throughout OS's and works the same way everywhere then the keypress playback could also be used, although it would be still a top level script-kiddie hackery... :(

**Comment by Andy2244** [comment]
I would prefer a way that might also work with upcoming windows versions, instead of a ahk script or simulating key presses.

Why not add this as admin only option? I always use a admin account anyway so the option could explain that you need admin rights and preferable be grayed out if you don't started it as admin or from a admin account. This way it should become clear why this feature currently does not work.

**Comment by sirWest** [comment]
This is not a bad idea at all, I'll start looking into it :)

**Comment by sirWest** [comment]
now, I MUST warn you, this is so alpha that it's basically just a pile of ****.
https://drive.google.com/file/d/0B8VyHhqTufh1M0hJaS1VNlpXWXc/view?usp=sharing
If you dare give it a go. The speaker config is located on the device specific tab in settings window, enabled only when ran as admin. Since I don't possess any soundcard with any other than either 2.0 or SPDIF output then I just don't have even anything to test it with. Good luck! :P

**Comment by sirWest** [comment]
I know that his version doesn't remember the device if it's uninstalled and then installed again, please just test the speaker setting change if it works. It should work already when changing this setting in the settings window when changing the combobox value. But, BUT, I now found a way to really remember a device by it's combined hardware ID and output-specific GUID :dancers: For that I must get driver ID from WMI, then go read registry and then read MMDevice ID's and combining all that mambo-jumbo I can actually remember a device even if you put it to another port or reinstall the driver completely! This is a insane and crazy logic but right now it's working with my USB soundcard and I can kinda confirm it on my side. I hope I can give another version sometime during the next 12 hours. Your feedback of the version given would still be more than helpful!

**Comment by Andy2244** [comment]
thx will test this later today

**Comment by Andy2244** [comment]
Tried the 2.1 alpha and did not work.

1) Did run it on a admin account and used the "Run as Admin" option as well.

2) The enumeration list of available options (stereo to 7.1) seems to query the currently/last set windows speaker configuration. This means if windows is set to stereo the list wont even allow to pick anything higher and only direct/stereo is available. So you need to query something else or simply assume the full list similar to what windows offers as options (stereo/5.1/7.1)

3) If i set windows to 7.1 and in AS pick 5.1/stereo and try change the device back and forth nothing actually changes, windows always keeps what was set via the default sound/configuration dialog.

So something is not working.

PS: Here is the setting i see:

```
<Device DeviceID="{0.0.0.00000000}.{bc63864f-48eb-4577-b249-1972cab784e5}" HideFromList="false" Hue="0" Saturation="0" Brightness="0" SpeakerConfig="KSAUDIO_SPEAKER_STEREO" SetSpeakerConfig="true" />
```

**Comment by sirWest** [comment]
That's really sad to hear :( The only way for me to try and understand is when you'd help me on this a little bit. Here's what I'd ask you to do, it you're up for it:
In regedit there's a key named HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\MMDevices\Audio\Render
Set your speaker config for this surround device to 2.0, then right-click that whole registry folder and choose export. Export it to a file, name it something that says it's with 2.0 settings. Now switch to 5.1, don't change ANYTHING else for any audio devices, press F5 for refreshing the registry and then export that same key again, naming it something 5.1 related.
As far as I know these settings don't contain anything sensitive to you or your system, so you can just pack them to a zip file and post it here or somewhere. I'd like to see what changes there, and IF anything changes at all. I just plugged my TV to my laptop but the speaker configuration shows only regular stereo for me, so I just can't simply try these things myself at all, it's very depressing to test/develop like this...

**Comment by Andy2244** [comment]
Does your onboard soundcard only have stereo capabilities and don't have the usual 3x audio outputs on the back-plate, which indicates its 5.1/7.1 capable?

At least my Xi-Fi and the Realtek all list 5.1/7.1 option if i check the CP->Sound->device(speaker)->configuration option.

PS: If still needed can export those reg settings later.

**Comment by sirWest** [comment]
Sadly not, I have a 3 year old Fujitsu laptop :(

**Comment by Andy2244** [comment]
Here are the regs, they show distinct hex differences for all 3 modes.
https://drive.google.com/file/d/0B-4v-tWdPLgCeU9TY3hFRmJDRVU/view?usp=sharing

**Comment by sirWest** [comment]
thanks, this is really useful info. This shows that in reality the config changes also the PKEY_AudioEngine_DeviceFormat key value, which must be changed by the IDeviceTopology interface. I currently have a pile of different CoreAudio C# libraries downloaded, I have to add some of the stuff that I have created myself and combine them to a whole huge perfect library and THEN I should be able to change this correctly. I will let you know here if I have anything to give for testing :) And a BIG Thank You for your help!!!

**Comment by sirWest** [comment]
I've discovered some more things using the IDeviceTopology interface where you can actually set the channel configuration etc. This needs to be supposedly used together with the way that I was already doing it. Sad to say that unless I have an actual surround sound card then I just can't develop this. That's why I took the money that is donated to me and ordered this: http://www.aliexpress.com/item/Multifunction-7-1-Channel-external-5-1-Sound-Card-With-Optical-For-Laptop-Computer-5-1ch/32219646667.html
Now I really hope that it works in the true way a surround sound card has to work, because if it doesn't then I'm out of ideas and options. Since this comes from China and I live in Estonia (Europe) then it might take up to a month before I get this. So in the meantime I'll just gather info and ideas and really hope this all pays off as a working result :)

**Comment by Andy2244** [comment]
thx for taking this much effort, its appreciated!

**Comment by Andy2244** [comment]
Quick feedback from my Windows 10 (10166) machine:

It seems that now Windows at least remembers the 5.1 settings if you sleep/hibernate, no matter what state your AVR was in. So if the AVR is off, windows will list the device as "stereo" only and if you turn the AVR on it will switch back to the last selected surround mode.

However if you restart Win10 while the AVR is off than windows wont remember the setting and use stereo even after the AVR was switched on.

So basically we still loose our speaker configuration every time we reboot windows and the AVR is off, while doing so.

bye
Andy

**Comment by sirWest** [comment]
I have studied the options for this for days during last few months and I've come to a conclusion that it's basically impossible to do. Even though running as admin would help with access, the whole speaker configuration settings part in the registry is just so damn puzzling and complicated, besides that it's also different for different sound device manufacturer. The final piece that makes me give up on it is that some sound device manufacturers don't even use those settings and have some custom functionality in their drivers to switch and configure speakers: that's because using windows' own logic is not even mandatory! So, sad to say but I just can't see it happening, unless some real specialist from either Microsoft or someone from some sound device manufacturer comes and helps us out... I'm really sorry :(
